lib/rubocop/cop/style/redundant_line_continuation.rb

```diff
@@ -0,0 +1,147 @@
+      # for string concatenation is not redundant and is not considered an offense.
+      # And if removing the backslash changes the behavior, it will not be considered an offense.
+      #
+      # @example
+      #   # bad
+      #   foo. \
+      #     bar
+      #   foo \
+      #     &.bar \
+      #       .baz
+      #
+      #   # good
+      #   foo.
+      #     bar
+      #   foo
+      #     &.bar
+      #       .baz
+      #
+      #   # bad
+      #   [foo, \
+      #     bar]
+      #   {foo: \
+      #     bar}
+      #
+      #   # good
+      #   [foo,
+      #     bar]
+      #   {foo:
+      #     bar}
+      #
+      #   # bad
+      #   foo(bar, \
+      #     baz)
+      #
+      #   # good
+      #   foo(bar,
+      #     baz)
+      #
+      #   # also good - backslash in string concatenation is not redundant
+      #   foo('bar' \
+      #     'baz')
+      #
+      #   # also good - backslash at the end of a comment is not redundant
+      #   foo(bar, # \
+      #     baz)
+      #
+      #   # also good - backslash at the line following the newline begins with a + or -,
+      #   # it is not redundant
+      #   1 \
+      #     + 2 \
+      #       - 3
+      #
+      #   # also good - backslash with newline between the method name and its arguments,
+      #   # it is not redundant.
+      #   some_method \
+      #     (argument)
+      #
+      class RedundantLineContinuation < Base
+        include MatchRange
+        include RangeHelp
+        extend AutoCorrector
+
+        MSG = 'Redundant line continuation.'
+        STR_TYPES = %i[str dstr xstr].freeze
+
+        def on_new_investigation
+          return unless processed_source.ast
+
+          each_match_range(processed_source.ast.source_range, /(\\\n)/) do |range|
+            next if not_redundant?(range)
+            next unless valid_syntax?(range)
+
+            add_offense(range) do |corrector|
+              corrector.remove_leading(range, 1)
+            end
+          end
+        end
+
+        private
+
+        def not_redundant?(range)
+          !ends_with_backslash_without_comment?(range.source_line) ||
+            string_concatenation?(range.source_line) ||
+            starts_with_plus_or_minus?(processed_source[range.line])
+        end
+
+        def ends_with_backslash_without_comment?(source_line)
+          source_line.gsub(/#.+/, '').end_with?('\\')
+        end
+
+        def string_concatenation?(source_line)
+          /["']\s*\\\z/.match?(source_line)
+        end
+
+        def valid_syntax?(range)
+          return true unless (node = find_node_for_line(range.line))
+          return false if argument_newline?(node)
+
+          ProcessedSource.new(node.source.gsub(/\\\n/, "\n"),
+                              processed_source.ruby_version).valid_syntax?
+        end
+
+        def argument_newline?(node)
+          return unless node.send_type?
+
+          /#{node.method_name}\s*\\\n/.match?(node.source)
```

Comment:
AST should be preferred over regular expressions. How is it?

```suggestion
          return if !node.send_type? || node.arguments.empty?

          node.loc.selector.line != node.first_argument.loc.line
```